geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureCollection.scala

```diff
@@ -11,96 +11,248 @@ package org.locationtech.geomesa.index.geotools
+    o match {
+      case f: SimpleFeature =>
+        val sub = subCollection(FilterHelper.ff.id(f.getIdentifier))
+        WithClose(CloseableIterator(sub.features()))(_.nonEmpty)
+
+      case _ => false
+    }
+  }
+
+  override def containsAll(collection: java.util.Collection[_]): Boolean = {
+    val size = collection.size()
+    if (size == 0) { true } else {
+      val filters = Seq.newBuilder[Filter]
+      filters.sizeHint(size)
+
+      val features = collection.iterator()
+      while (features.hasNext) {
+        features.next match {
+          case f: SimpleFeature => filters += FilterHelper.ff.id(f.getIdentifier)
+          case _ => return false
+        }
+      }
+
+      val sub = subCollection(org.locationtech.geomesa.filter.orFilters(filters.result))
+      WithClose(CloseableIterator(sub.features()))(_.length) == size
+    }
+  }
 }
 
-object GeoMesaFeatureCollection {
+object GeoMesaFeatureCollection extends LazyLogging {
+
   private val oneUp = new AtomicLong(0)
+
   def nextId: String = s"GeoMesaFeatureCollection-${oneUp.getAndIncrement()}"
+
+  /**
+    * Attempts to visit the feature collection in an optimized manner. This will unwrap any decorating
+    * feature collections that may interfere with the `accepts` method.
+    *
+    * Note that generally this may not be a good idea - collections are presumably wrapped for a reason.
+    * However, our visitation functionality keeps being broken by changes in GeoServer, so we're being
+    * defensive here.
+    *
+    * @param collection feature collection
+    * @param visitor visitor
+    * @param progress progress monitor
+    */
+  def visit(
+      collection: FeatureCollection[SimpleFeatureType, SimpleFeature],
+      visitor: FeatureVisitor,
+      progress: ProgressListener = new NullProgressListener): Unit = {
+    val unwrapped = if (collection.isInstanceOf[GeoMesaFeatureVisitingCollection]) { collection } else {
+      try { unwrap(collection) } catch {
+        case e: Throwable => logger.debug("Error trying to unwrap feature collection:", e); collection
+      }
+    }
+    unwrapped.accepts(visitor, progress)
+  }
+
+  /**
+    * Attempts to remove any decorating feature collections
+    *
+    * @param collection collection
+    * @param level level of collections that have been removed, used to detect potentially infinite looping
+    * @return
+    */
+  @tailrec
+  private def unwrap(
+      collection: FeatureCollection[SimpleFeatureType, SimpleFeature],
+      level: Int = 1): FeatureCollection[SimpleFeatureType, SimpleFeature] = {
+
+    // noinspection TypeCheckCanBeMatch
+    val unwrapped = if (collection.isInstanceOf[DecoratingSimpleFeatureCollection]) {
+      getDelegate(collection, classOf[DecoratingSimpleFeatureCollection])
+    } else if (collection.isInstanceOf[DecoratingFeatureCollection[SimpleFeatureType, SimpleFeature]]) {
+      getDelegate(collection, classOf[DecoratingFeatureCollection[SimpleFeatureType, SimpleFeature]])
```

Comment:
Does GeoServer have a separate decorating interface that they could be using?  Aka, do we need to do anything extreme to cover another corner case?

Reply:
I didn't find anything... I looked through all the classes that I could find that extend SimpleFeatureCollection or FeatureCollection. They all either delegate appropriately, or extend one of these two.